### PR TITLE
ABI: Only store bitwise take-able values inline

### PR DESCRIPTION
**Pull request by aschwaighofer**
SR-343
rdar://31414907

### CODE REVIEW

**Comment by aschwaighofer** [comment]
@swift-ci Please test

**Comment by aschwaighofer** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
Thanks!

File: lib/IRGen/GenValueWitness.cpp [hunk 1]
```diff
@@ -1288,6 +1291,10 @@ FixedPacking TypeInfo::getFixedPacking(IRGenModule &IGM) const {
   if (!fixedTI)
     return FixedPacking::Dynamic;
 
+  // By convention we don't store bitwise takable values inline.
```
Comment (slavapestov) [inline lib/IRGen/GenValueWitness.cpp]: Is this comment wrong? We don't store *non*-bitwise takable values inline right?

**Comment by slavapestov** [comment]
What about resilience, where a value might be bitwise takable in one resilience domain but not in another?

**Comment by slavapestov** [comment]
I'm assuming you can now optimize value witness functions for existentials to memcpy the inline buffer. Is that correct?

Also more tests would be nice. In particular I'm worried about resilience here. If two different modules have a different idea of whether a struct is bitwise takable, will we generate invalid code if one module stores a struct into an existential and another one loads from it?

**Comment by jckarter** [comment]
Determining whether a resilient type can be stored inline in an existential buffer is already a dynamic judgment, since you don't know the size, so we should ask the type metadata the question dynamically. I agree it's worth adding tests for.

File: lib/IRGen/GenValueWitness.cpp (hunk 1 above)
Comment (aschwaighofer) [reply lib/IRGen/GenValueWitness.cpp]: Yes. The comment is inverted.

**Comment by aschwaighofer** [comment]
@slavapestov 
> What about resilience, where a value might be bitwise takable in one resilience domain but not in another?

 If we access a value resiliently than we also have to accesss bitwise takability/isInline resiliently.
There is an existing test case that verifies that the isInline flag has changed.

I will add some more tests.

**Comment by aschwaighofer** [comment]
@slavapestov 
> I'm assuming you can now optimize value witness functions for existentials to memcpy the inline buffer. 

Correct. I can now remove the existential value witness: initializeBufferWithTakeOfBuffer (i.e a move of an existential) and replace it with calls to memcpy. 
Either, the value in the existential buffer is inline and bitwise takable in which case a memcpy is safe, or the value in the existential buffer is a reference to the heap object holding the value in which case a memcpy is also safe.

**Comment by aschwaighofer** [comment]
@swift-ci Please test

**Comment by aschwaighofer** [comment]
@swift-ci Please test

**Pull request merged**